packages/massa-web3/src/experimental/smartContract.ts

```diff
@@ -207,13 +239,29 @@ export class SmartContract {
   }
 
   /**
-   * Estimates the gas required for an operation.
-   * Currently, it returns a predefined maximum gas value.
-   * @returns A promise that resolves to the estimated gas amount in bigint.
-   * TODO: Implement dynamic gas estimation using dry run call.
-   * TODO: Remove static if needed.
+   * Returns the gas estimation for a given function.
+   *
+   * @param func - The function to estimate the gas cost.
+   * @returns The gas estimation for the function.
    */
-  static async getGasEstimation(): Promise<bigint> {
-    return MAX_GAS_CALL
+  async getGasEstimation(
+    func: string,
+    parameter: Uint8Array,
+    callerAddress: Address,
+    opts: CallOptions
+  ): Promise<bigint> {
+    const result = await this.read(func, {
+      parameter: parameter,
+      maxGas: opts.maxGas,
+      fee: opts.fee,
+      callerAddress: callerAddress.toString(),
+      coins: opts.coins,
+    })
+
+    if (result.info.error) {
+      throw new Error(result.info.error)
+    }
+
+    return BigInt(result.info.gasCost)
```

Comment:
Aren't we supposed to have a security marging?

Reply:
Not sure as it was not the case before but I'll ask